compile/cycle_test.go

```diff
@@ -0,0 +1,157 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compile
+
+import (
+	"testing"
+
+	"github.com/thriftrw/thriftrw-go/ast"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// Helper to create typedefs with cyclic references inside the test table.
+func withTypedef(f func(t *TypedefSpec)) TypeSpec {
+	t := &TypedefSpec{File: "test.thrift"}
+	f(t)
+	return t
+}
+
+func TestFindTypeCycles(t *testing.T) {
+	tests := []struct {
+		desc string
+		typ  TypeSpec
+		msgs []string
+	}{
+		{
+			desc: "self-referential typedef",
+			typ: withTypedef(func(t *TypedefSpec) {
+				t.Name = "foo"
+				t.Target = t
+			}),
+			msgs: []string{
+				"found a type reference cycle",
+				"   foo",
+				"-> foo",
+			},
+		},
+		{
+			desc: "mutually recursive references",
+			typ: withTypedef(func(x *TypedefSpec) {
+				x.Name = "foo"
+				x.Target = &TypedefSpec{
+					Name:   "bar",
+					File:   "test.thrift",
+					Target: x,
+				}
+			}),
+			msgs: []string{
+				"found a type reference cycle",
+				"   foo",
+				"-> bar",
+				"-> foo",
+			},
+		},
+		{
+			desc: "recurse from map",
+			typ: withTypedef(func(t *TypedefSpec) {
+				t.Name = "foo"
+				t.Target = &MapSpec{
+					KeySpec:   StringSpec,
+					ValueSpec: t,
+				}
+			}),
+			msgs: []string{
+				"found a type reference cycle",
+				"   foo",
+				"-> map<string, foo>",
+				"-> foo",
+			},
+		},
+		{
+			desc: "recurse from list",
+			typ: withTypedef(func(t *TypedefSpec) {
+				t.Name = "foo"
+				t.Target = &ListSpec{ValueSpec: t}
+			}),
+			msgs: []string{
+				"found a type reference cycle",
+				"   foo",
+				"-> list<foo>",
+				"-> foo",
+			},
+		},
+		{
+			desc: "recurse from set",
+			typ: withTypedef(func(t *TypedefSpec) {
+				t.Name = "foo"
+				t.Target = &SetSpec{ValueSpec: t}
+			}),
+			msgs: []string{
+				"found a type reference cycle",
+				"   foo",
+				"-> set<foo>",
+				"-> foo",
+			},
+		},
+		{
+			desc: "recurse from struct",
+			typ: withTypedef(func(t *TypedefSpec) {
+				t.Name = "foo"
+				t.Target = &StructSpec{
+					Name: "bar",
+					File: "test.thrift",
+					Type: ast.StructType,
+					Fields: FieldGroup{
+						"foo": {
+							ID:       1,
+							Name:     "foo",
+							Type:     t,
+							Required: true,
+						},
+					},
+				}
+			}),
+			msgs: []string{
+				"found a type reference cycle",
+				"   foo",
+				"-> bar",
+				"-> foo",
+			},
+		},
+		// TODO test for cycles from structs
+	}
+
+	for _, tt := range tests {
+		typ := mustLink(t, tt.typ, scope())
+
+		err := findTypeCycles(typ)
+		if len(tt.msgs) > 0 {
```

Comment:
since you always have messages, feel free to remove the if, and assume it should always error.